pkg/util/deployment/deployment.go

```diff
@@ -367,6 +374,19 @@ func NewRSNewReplicas(deployment *extensions.Deployment, allRSs []*extensions.Re
 	}
 }
 
+// IsSaturated checks if the new replica set is saturated by comparing its size with its deployment size.
+func IsSaturated(deployment *extensions.Deployment, rs *extensions.ReplicaSet) bool {
+	if rs == nil {
+		return false
+	}
+	desiredString := rs.Annotations[DesiredReplicasAnnotation]
+	desired, err := strconv.Atoi(desiredString)
+	if err != nil {
+		return false
+	}
+	return rs.Spec.Replicas == deployment.Spec.Replicas && int32(desired) == deployment.Spec.Replicas
```

Comment:
A replica set is saturated when d.Spec.Replicas == rs.Spec.Replicas but you cannot be sure when you proportionally scale that this holds true prior to scaling a replica set. For example:

deployment:
nginx.Spec.Replicas == 10 (maxSurge: 1, revision: 2)

replica sets:
nginx-1.Spec.Replicas == 5 (desired-replicas: 10, total-replicas: 11)
nginx-2.Spec.Replicas == 5 (desired-replicas: 10, total-replicas: 11)

We are in the middle of a rollout (paused or not it doesn't matter because we will scale proportionally in both cases)

User or autoscaler scales the nginx deployment to 5 replicas.

nginx.Spec.Replicas == 5

Without the int32(desired) == deployment.Spec.Replicas check, the controller would see:

nginx-2.Spec.Replicas == nginx.Spec.Replicas == 5

Identifies the new replica set as saturated and instantly scales down the old replica set (which we don't want to happen since that's is the reason we are doing proportional scaling in the first place).

With the annotation check you have:

5 == 5 && 10 == 5 which will avoid to identify the new replica set as saturated thus continue proportional scaling.